tenant/duplicate_reads.go

```diff
@@ -0,0 +1,362 @@
+package tenant
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/influxdata/influxdb"
+	"github.com/influxdata/influxdb/kv"
+	"go.uber.org/zap"
+)
+
+var _ influxdb.BucketService = (*dupReadBucketService)(nil)
+var _ influxdb.OrganizationService = (*dupReadOrganizationService)(nil)
+var _ influxdb.UserResourceMappingService = (*dupReadUserResourceMappingService)(nil)
+var _ influxdb.UserService = (*dupReadUserService)(nil)
+
+// readOnlyStore is a wrapper for kv.Store that ensures that updates are not applied.
+type readOnlyStore struct {
+	kv.Store
+	log *zap.Logger
+}
+
+func (r readOnlyStore) Update(context.Context, func(kv.Tx) error) error {
+	r.log.Warn("attempted update on read-only store")
+	return nil
+}
+
+type tenantService struct {
+	influxdb.BucketService
+	influxdb.OrganizationService
+	influxdb.UserResourceMappingService
+	influxdb.UserService
+	influxdb.PasswordsService
+}
+
+// NewReadOnlyTenantService returns a influxdb.TenantService that cannot update the underlying store.
+func NewReadOnlyTenantService(log *zap.Logger, store kv.Store) (influxdb.TenantService, error) {
+	ro, err := NewStore(readOnlyStore{
+		Store: store,
+		log:   log,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return NewService(ro), nil
+}
+
+// NewDuplicateReadTenantService returns a tenant service that duplicates the reads of a kv.Service to a tenant.Service.
+// It does so by creating a tenant.Service with the given store.
+// For consistent results, the given kv.Service should be built on top of the same store.
+// The foreseen use case is to compare two service versions, an old one and a new one.
+// The resulting influxdb.TenantService:
+//  - forwards writes to the old service;
+//  - reads from the old one, if no error is encountered, it reads from the new one;
+//  - compares the results obtained and logs the difference, if any.
+func NewDuplicateReadTenantService(log *zap.Logger, oldSvc *kv.Service, store kv.Store) (influxdb.TenantService, error) {
+	// Using a read-only service to further ensure that no update is applied to the underlying store.
+	svc, err := NewReadOnlyTenantService(log, store)
+	if err != nil {
+		return nil, err
+	}
+	return tenantService{
+		BucketService:              NewDuplicateReadBucketService(log, oldSvc, svc),
+		OrganizationService:        NewDuplicateReadOrganizationService(log, oldSvc, svc),
+		UserResourceMappingService: NewDuplicateReadUserResourceMappingService(log, oldSvc, svc),
+		UserService:                NewDuplicateReadUserService(log, oldSvc, svc),
+		PasswordsService:           NewDuplicateReadPasswordService(log, oldSvc, svc),
+	}, nil
```

Comment:
I think the functions could just be on the `tenantService` instead of making separate structs for each sub system. When we rollin feature flagging I wouldn't want all the sub structs to have to share information about who or what systems are going to write. I suspect it would be better to flatten this.